chain/epoch_manager/src/validator_selection.rs

```diff
@@ -487,6 +488,145 @@ mod tests {
+    ) {
+        let epoch_config = create_epoch_config(
+            num_shards,
+            num_block_producer_seats,
+            0,
+            ValidatorSelectionConfig {
+                num_chunk_only_producer_seats: 0,
+                minimum_validators_per_shard,
+                minimum_stake_ratio: minimum_stake_ratio,
+            },
+        );
+
+        let prev_epoch_height = 18;
+        let prev_epoch_info = create_prev_epoch_info(prev_epoch_height, &["test1", "test2"], &[]);
+        let proposals = create_proposals((1..=200).flat_map(|i| {
+            // Each shard gets a pair of validators, one with twice as
+            // much stake as the other.
+            vec![(format!("test{}", i), 1000), (format!("test{}", 200 + i), i * 1000)].into_iter()
+        }));
+        let epoch_info = proposals_to_epoch_info(
+            &epoch_config,
+            [0; 32],
+            &prev_epoch_info,
+            proposals,
+            Default::default(),
+            Default::default(),
+            1000,
+            PROTOCOL_VERSION,
+        )
+        .unwrap();
+
+        // make sure, that all chunk producers are actually block producers
+        println!("num_shards = {}", num_shards);
+        let mut counts: Vec<usize> = vec![0; num_block_producer_seats as usize];
+
+        for shard_id in 0..num_shards {
+            let chunk = &epoch_info.chunk_producers_settlement()[shard_id as usize];
+            assert!(chunk.len() as u64 >= minimum_validators_per_shard);
+            for validator_id in chunk.iter() {
+                counts[*validator_id as usize] += 1;
+            }
+        }
+
+        let mut prev_stake = epoch_info.get_validator(0).stake();
+        let min_chunks_per_validator =
+            ((num_shards * minimum_validators_per_shard) / num_block_producer_seats) as usize;
+        let max_chunks_per_validator = ((num_shards * minimum_validators_per_shard - 1)
+            / num_block_producer_seats
+            + 1) as usize;
+        for validator_id in 0..num_block_producer_seats {
+            let v = epoch_info.get_validator(validator_id);
+            // Make sure, stakes are in decreasing order
+            assert!(prev_stake >= v.stake());
+            assert!(counts[validator_id as usize] >= min_chunks_per_validator);
+            assert!(counts[validator_id as usize] <= max_chunks_per_validator);
+            println!(
+                "      validator_id = {} account = {} stacked = {} in {} chunks",
+                validator_id,
+                v.account_id(),
+                v.stake(),
+                counts[validator_id as usize]
+            );
+            prev_stake = v.stake();
+        }
+
+        assert_eq!(epoch_info.block_producers_settlement().len() as u64, num_block_producer_seats);
+        assert_eq!(epoch_info.chunk_producers_settlement().len() as u64, num_shards);
+    }
+
+    #[test]
+    fn test_block_producer_sampling_with_100_block_producers() {
```

Comment:
Probably makes sense to write something that can take some sort of config as input (similar to how runtime-tester is implemented) to make writing tests easier

Reply:
Done :)